# Aktivitas Partisipatif #1

### 1. MEMANGGIL DATA

### Dataset diambil dari Google Drive menggunakan library pandas. Google Drive diakses melalui Colab agar file bisa dibaca. Dataset dimuat dengan fungsi pd.read_excel().

In [ ]:
import pandas as pd
from google.colab import drive

drive.mount('/content/drive')
file_path = '/content/drive/MyDrive/Colab Notebooks/CAD alizadeh.xls'
data = pd.read_excel(file_path, sheet_name='CAD')
data.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


No  Age  Weight  Length    Sex        BMI  DM  HTN  Current Smoker  \
0   1   53      90     175   Male  29.387755   0    1               1   
1   2   67      70     157  Fmale  28.398718   0    1               0   
2   3   54      54     164   Male  20.077335   0    0               1   
3   4   66      67     158  Fmale  26.838648   0    1               0   
4   5   50      87     153  Fmale  37.165193   0    1               0   

   EX-Smoker  ...    K   Na    WBC Lymph Neut  PLT EF-TTE Region RWMA     VHD  \
0          0  ...  4.7  141   5700    39   52  261     50           0       N   
1          0  ...  4.7  156   7700    38   55  165     40           4       N   
2          0  ...  4.7  139   7400    38   60  230     40           2    mild   
3          0  ...  4.4  142  13000    18   72  742     55           0  Severe   
4          0  ...  4.0  140   9200    55   39  274     50           0  Severe   

     Cath  
0     Cad  
1     Cad  
2     Cad  
3  Normal  
4  Normal  

[5 rows x 56 columns]

### 2. MENAMPILKAN DATA

### Menampilkan informasi dasar dataset seperti jumlah baris, kolom, dan tipe data. Juga menampilkan statistik deskriptif seperti rata-rata, nilai minimum, dan maksimum.

In [ ]:
# Menampilkan informasi dasar dataset
print("Informasi dataset:")
data.info()

# Menampilkan statistik deskriptif
print("\nStatistik deskriptif:")
data.describe()

Informasi dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 56 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   No                     303 non-null    int64  
 1   Age                    303 non-null    int64  
 2   Weight                 303 non-null    int64  
 3   Length                 303 non-null    int64  
 4   Sex                    303 non-null    object 
 5   BMI                    303 non-null    float64
 6   DM                     303 non-null    int64  
 7   HTN                    303 non-null    int64  
 8   Current Smoker         303 non-null    int64  
 9   EX-Smoker              303 non-null    int64  
 10  FH                     303 non-null    int64  
 11  Obesity                303 non-null    object 
 12  CRF                    303 non-null    object 
 13  CVA                    303 non-null    object 
 14  Airway disease         303 non-null    

No         Age      Weight      Length         BMI          DM  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean   152.000000   58.897690   73.831683  164.716172   27.248339    0.297030   
std     87.612784   10.392278   11.987358    9.327661    4.098865    0.457706   
min      1.000000   30.000000   48.000000  140.000000   18.115413    0.000000   
25%     76.500000   51.000000   65.000000  158.000000   24.514380    0.000000   
50%    152.000000   58.000000   74.000000  165.000000   26.775510    0.000000   
75%    227.500000   66.000000   81.000000  171.000000   29.411765    1.000000   
max    303.000000   86.000000  120.000000  188.000000   40.900658    1.000000   

              HTN  Current Smoker   EX-Smoker          FH  ...         ESR  \
count  303.000000      303.000000  303.000000  303.000000  ...  303.000000   
mean     0.590759        0.207921    0.033003    0.158416  ...   19.462046   
std      0.492507        0.406491    0.178941    0.365734  ...   15.936475   
min      0.000000        0.000000    0.000000    0.000000  ...    1.000000   
25%      0.000000        0.000000    0.000000    0.000000  ...    9.000000   
50%      1.000000        0.000000    0.000000    0.000000  ...   15.000000   
75%      1.000000        0.000000    0.000000    0.000000  ...   26.000000   
max      1.000000        1.000000    1.000000    1.000000  ...   90.000000   

               HB           K          Na           WBC       Lymph  \
count  303.000000  303.000000  303.000000    303.000000  303.000000   
mean    13.153465    4.230693  140.996700   7562.046205   32.399340   
std      1.610452    0.458202    3.807885   2413.739323    9.972592   
min      8.900000    3.000000  128.000000   3700.000000    7.000000   
25%     12.200000    3.900000  139.000000   5800.000000   26.000000   
50%     13.200000    4.200000  141.000000   7100.000000   32.000000   
75%     14.200000    4.500000  143.000000   8800.000000   39.000000   
max     17.600000    6.600000  156.000000  18000.000000   60.000000   

             Neut         PLT      EF-TTE  Region RWMA  
count  303.000000  303.000000  303.000000   303.000000  
mean    60.148515  221.488449   47.231023     0.620462  
std     10.182493   60.796199    8.927194     1.132531  
min     32.000000   25.000000   15.000000     0.000000  
25%     52.500000  183.500000   45.000000     0.000000  
50%     60.000000  210.000000   50.000000     0.000000  
75%     67.000000  250.000000   55.000000     1.000000  
max     89.000000  742.000000   60.000000     4.000000  

[8 rows x 35 columns]

### 3. Preprocessing data : memeriksa apakah ada yang kosong

### Memeriksa apakah ada data yang hilang (missing values) di dataset. Data kosong bisa mengganggu analisis, jadi perlu dicek.

In [ ]:
# Memeriksa missing values
missing_values = data.isnull().sum()
print("Missing Values:\n", missing_values)

Missing Values:
 No                       0
Age                      0
Weight                   0
Length                   0
Sex                      0
BMI                      0
DM                       0
HTN                      0
Current Smoker           0
EX-Smoker                0
FH                       0
Obesity                  0
CRF                      0
CVA                      0
Airway disease           0
Thyroid Disease          0
CHF                      0
DLP                      0
BP                       0
PR                       0
Edema                    0
Weak Peripheral Pulse    0
Lung rales               0
Systolic Murmur          0
Diastolic Murmur         0
Typical Chest Pain       0
Dyspnea                  0
Function Class           0
Atypical                 0
Nonanginal               0
Exertional CP            0
LowTH Ang                0
Q Wave                   0
St Elevation             0
St Depression            0
Tinversion               0
LVH        

### 4. Mengubah semua kolom menjadi bernilai numerik

### Algoritma machine learning butuh data numerik. Kolom kategorikal (seperti jenis kelamin) diubah ke numerik menggunakan one-hot encoding dengan pd.get_dummies().

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Buat salinan data untuk menghindari perubahan pada data asli
data_encoded = data.copy()

# Loop melalui setiap kolom dalam dataset
for column in data_encoded.columns:
    # Jika kolom bertipe object (kategorikal), lakukan Label Encoding
    if data_encoded[column].dtype == 'object':
        encoder = LabelEncoder()
        data_encoded[column] = encoder.fit_transform(data_encoded[column])

# Tampilkan data setelah encoding
print("\nData setelah semua kolom diubah menjadi numerik:")
data_encoded.head()


Data setelah semua kolom diubah menjadi numerik:


No  Age  Weight  Length  Sex        BMI  DM  HTN  Current Smoker  \
0   1   53      90     175    1  29.387755   0    1               1   
1   2   67      70     157    0  28.398718   0    1               0   
2   3   54      54     164    1  20.077335   0    0               1   
3   4   66      67     158    0  26.838648   0    1               0   
4   5   50      87     153    0  37.165193   0    1               0   

   EX-Smoker  ...    K   Na    WBC  Lymph  Neut  PLT  EF-TTE  Region RWMA  \
0          0  ...  4.7  141   5700     39    52  261      50            0   
1          0  ...  4.7  156   7700     38    55  165      40            4   
2          0  ...  4.7  139   7400     38    60  230      40            2   
3          0  ...  4.4  142  13000     18    72  742      55            0   
4          0  ...  4.0  140   9200     55    39  274      50            0   

   VHD  Cath  
0    1     0  
1    1     0  
2    3     0  
3    2     1  
4    2     1  

[5 rows x 56 columns]

### 5. Melakukan proses normalisasi pada masing-masing atribut

### Normalisasi mengubah nilai data ke rentang 0 sampai 1 agar semua fitur punya skala yang sama. Ini penting untuk algoritma yang sensitif terhadap perbedaan skala.

In [ ]:
from sklearn.preprocessing import MinMaxScaler

# Normalisasi data setelah semua kolom diubah menjadi numerik
scaler = MinMaxScaler()
data_normalized = pd.DataFrame(scaler.fit_transform(data_encoded), columns=data_encoded.columns)

# Tampilkan data setelah normalisasi
print("\nData setelah preprocessing dan normalisasi:")
data_normalized.head()


Data setelah preprocessing dan normalisasi:


No       Age    Weight    Length  Sex       BMI   DM  HTN  \
0  0.000000  0.410714  0.583333  0.729167  1.0  0.494721  0.0  1.0   
1  0.003311  0.660714  0.305556  0.354167  0.0  0.451314  0.0  1.0   
2  0.006623  0.428571  0.083333  0.500000  1.0  0.086105  0.0  0.0   
3  0.009934  0.642857  0.263889  0.375000  0.0  0.382846  0.0  1.0   
4  0.013245  0.357143  0.541667  0.270833  0.0  0.836058  0.0  1.0   

   Current Smoker  EX-Smoker  ...         K        Na       WBC     Lymph  \
0             1.0        0.0  ...  0.472222  0.464286  0.139860  0.603774   
1             0.0        0.0  ...  0.472222  1.000000  0.279720  0.584906   
2             1.0        0.0  ...  0.472222  0.392857  0.258741  0.584906   
3             0.0        0.0  ...  0.388889  0.500000  0.650350  0.207547   
4             0.0        0.0  ...  0.277778  0.428571  0.384615  0.905660   

       Neut       PLT    EF-TTE  Region RWMA       VHD  Cath  
0  0.350877  0.329149  0.777778          0.0  0.333333   0.0  
1  0.403509  0.195258  0.555556          1.0  0.333333   0.0  
2  0.491228  0.285914  0.555556          0.5  1.000000   0.0  
3  0.701754  1.000000  0.888889          0.0  0.666667   1.0  
4  0.122807  0.347280  0.777778          0.0  0.666667   1.0  

[5 rows x 56 columns]

# Aktivitas Partisipatif #2

In [ ]:
#import library
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

### 6.  Membagi Data Menjadi Data Latih dan Data Uji

Setelah data siap, langkah selanjutnya adalah membagi dataset menjadi dua bagian:

Data Latih (Training Data): Digunakan untuk melatih model.

Data Uji (Testing Data): Digunakan untuk menguji seberapa baik model bekerja.

pakai fungsi train_test_split dari sklearn untuk membagi data.
Biasanya, 80% data digunakan untuk latih, dan 20% untuk uji.

In [ ]:
target_column = 'HTN'

# Membagi data menjadi 80% data latih dan 20% data uji
X = data_normalized.drop(target_column, axis=1)  # Fitur (semua kolom kecuali target)
y = data_normalized[target_column]  # Target (kolom 'Current Smoker')
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### 7. Membuat Model Naive Bayes

Buat model menggunakan algoritma Naive Bayes. Naive Bayes adalah salah satu algoritma klasifikasi yang sederhana tapi cukup efektif untuk masalah klasifikasi, terutama jika dataset tidak terlalu besar.

Pakai GaussianNB dari library sklearn untuk membuat model.
Model ini dilatih menggunakan data latih (X_train dan y_train).

In [ ]:
from sklearn.naive_bayes import GaussianNB

# Membuat model Naive Bayes
model = GaussianNB()

# Melatih model menggunakan data latih
model.fit(X_train, y_train)

# Memprediksi data testing
y_pred_nb = model.predict(X_test)

### 8.  Memprediksi dan Mengevaluasi Model

Setelah prediksi selesai, hitung akurasi model. Akurasi adalah persentase prediksi yang benar dibandingkan dengan total prediksi.

Hasil Akurasi:

Dalam kasus ini, akurasi model adalah 55.74%. Artinya, model bisa memprediksi dengan benar sekitar 55.74% dari data uji.

Confusion Matrix:

Confusion Matrix adalah tabel yang menunjukkan berapa banyak prediksi yang benar dan salah.

Misalnya:

True Positive (TP): Prediksi benar positif.

False Positive (FP): Prediksi salah positif.

True Negative (TN): Prediksi benar negatif.

False Negative (FN): Prediksi salah negatif.

Laporan Klasifikasi:

Laporan ini memberikan informasi lebih detail tentang performa model, seperti:

Precision: Seberapa akurat prediksi positif.

Recall: Seberapa banyak data positif yang berhasil diprediksi.

F1-Score: Gabungan dari precision dan recall.



In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# hitung akurasi
accuracy_nb = accuracy_score(y_test, y_pred_nb)
print(f"Akurasi: {accuracy_nb * 100:.2f}%")

# print confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred_nb)
print("Confusion Matrix:")
print(conf_matrix)

# print laporan klasifikasi
class_report = classification_report(y_test, y_pred_nb)
print("Laporan Klasifikasi:")
print(class_report)

Akurasi: 57.38%
Confusion Matrix:
[[23  2]
 [24 12]]
Laporan Klasifikasi:
              precision    recall  f1-score   support

         0.0       0.49      0.92      0.64        25
         1.0       0.86      0.33      0.48        36

    accuracy                           0.57        61
   macro avg       0.67      0.63      0.56        61
weighted avg       0.71      0.57      0.55        61



# Aktivitas Partisipatif #3

#####RANDOM FOREST
RandomForestClassifier digunakan untuk membuat model Random Forest, yang merupakan algoritma pembelajaran mesin berbasis ensemble. Selain itu, accuracy_score, confusion_matrix, dan classification_report digunakan untuk mengevaluasi kinerja model setelah dilatih dan diuji.

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

model Random Forest dibuat dengan 100 pohon keputusan (decision trees). Setiap pohon dalam Random Forest bekerja dengan cara yang sedikit berbeda, sehingga hasilnya lebih stabil dan mengurangi risiko overfitting. Parameter random_state=42 digunakan untuk memastikan bahwa hasil pelatihan model dapat direproduksi di lain waktu. Setelah itu, model dilatih dengan data latih yang sudah disiapkan, yaitu fitur (X_train) dan target (y_train).

In [ ]:
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

Menggunakan model untuk memprediksi data uji dan menghitung akurasi prediksi.

In [ ]:
y_pred_rf = rf_model.predict(X_test)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print("Akurasi Random Forest: {:.2f}%".format(accuracy_rf * 100))

Akurasi Random Forest: 78.69%



Menampilkan confusion matrix untuk melihat seberapa baik model mengklasifikasikan data.

In [ ]:
print("\nConfusion Matrix Random Forest:")
print(confusion_matrix(y_test, y_pred_rf))


Confusion Matrix Random Forest:
[[17  8]
 [ 5 31]]


Menampilkan laporan klasifikasi dengan metrik seperti precision, recall, dan F1-score.

In [ ]:
print("\nLaporan Klasifikasi Random Forest:")
print(classification_report(y_test, y_pred_rf))


Laporan Klasifikasi Random Forest:
              precision    recall  f1-score   support

         0.0       0.77      0.68      0.72        25
         1.0       0.79      0.86      0.83        36

    accuracy                           0.79        61
   macro avg       0.78      0.77      0.78        61
weighted avg       0.79      0.79      0.78        61



Berdasarkan perbandingan akurasi antara Naive Bayes dan Random Forest, terlihat bahwa Random Forest lebih unggul dengan akurasi mencapai 78.69%, sementara Naive Bayes hanya memperoleh 57.38%.


In [ ]:
print("\n=== PERBANDINGAN AKURASI ===")
print(f"Naive Bayes\t: {accuracy_nb:.4f} ({accuracy_nb*100:.2f}%)")
print(f"Random Forest\t: {accuracy_rf:.4f} ({accuracy_rf*100:.2f}%)")


=== PERBANDINGAN AKURASI ===
Naive Bayes	: 0.5738 (57.38%)
Random Forest	: 0.7869 (78.69%)


Random Forest lebih unggul karena dapat menangani hubungan kompleks antar fitur dan mengurangi overfitting dengan menggunakan banyak pohon keputusan. Sementara itu, Naive Bayes mengasumsikan fitur independen, yang membuatnya kurang efektif ketika ada korelasi antar fitur dalam dataset.

# AKTIVITAS PARTISIPATIF #4

In [ ]:
# Import library yang diperlukan
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Membuat model MLP
mlp_model = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=42)

# Melatih model dengan data latih
mlp_model.fit(X_train, y_train)

# Memprediksi data uji
y_pred_mlp = mlp_model.predict(X_test)

# Menghitung akurasi
accuracy_mlp = accuracy_score(y_test, y_pred_mlp)
print(f"Akurasi MLP: {accuracy_mlp * 100:.2f}%")

# Menampilkan confusion matrix
print("\nConfusion Matrix MLP:")
print(confusion_matrix(y_test, y_pred_mlp))

# Menampilkan laporan klasifikasi
print("\nLaporan Klasifikasi MLP:")
print(classification_report(y_test, y_pred_mlp))

Akurasi MLP: 73.77%

Confusion Matrix MLP:
[[16  9]
 [ 7 29]]

Laporan Klasifikasi MLP:
              precision    recall  f1-score   support

         0.0       0.70      0.64      0.67        25
         1.0       0.76      0.81      0.78        36

    accuracy                           0.74        61
   macro avg       0.73      0.72      0.73        61
weighted avg       0.74      0.74      0.74        61



In [ ]:
from sklearn.model_selection import GridSearchCV

# Definisi parameter grid untuk tuning
param_grid = {
    'hidden_layer_sizes': [(50,), (100,), (50, 50), (100, 50)],
    'activation': ['relu', 'tanh', 'logistic'],
    'solver': ['adam', 'sgd'],
    'alpha': [0.0001, 0.001, 0.01],
    'learning_rate': ['constant', 'adaptive']
}

# Membuat model MLP untuk GridSearchCV
mlp = MLPClassifier(max_iter=1000, random_state=42)

# Melakukan GridSearchCV
grid_search = GridSearchCV(mlp, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train, y_train)

# Menampilkan parameter terbaik
print("Parameter terbaik:", grid_search.best_params_)

# Membuat model dengan parameter terbaik
best_mlp_model = grid_search.best_estimator_

# Memprediksi data uji
y_pred_best_mlp = best_mlp_model.predict(X_test)

# Evaluasi model terbaik
accuracy_best_mlp = accuracy_score(y_test, y_pred_best_mlp)
print(f"Akurasi MLP setelah tuning: {accuracy_best_mlp * 100:.2f}%")

# Menampilkan laporan klasifikasi
print("\nLaporan Klasifikasi MLP Terbaik:")
print(classification_report(y_test, y_pred_best_mlp))

Parameter terbaik: {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (50, 50), 'learning_rate': 'constant', 'solver': 'sgd'}
Akurasi MLP setelah tuning: 72.13%

Laporan Klasifikasi MLP Terbaik:
              precision    recall  f1-score   support

         0.0       0.65      0.68      0.67        25
         1.0       0.77      0.75      0.76        36

    accuracy                           0.72        61
   macro avg       0.71      0.72      0.71        61
weighted avg       0.72      0.72      0.72        61



/usr/local/lib/python3.11/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
